### PR TITLE
Parse and compare agent manifests

### PR DESCRIPTION
**Pull request by narrieta**
Parse the agent manifests in vmSettings and compare them against extensionsConfig.

In the process, I removed VMAgentManifestUri, VMAgentManifestList, ExtHandlerVersionUri, and ExtHandlerPackageUri and replaced them with simple types (strings and lists)

### CODE REVIEW

File: azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py [hunk 1]
```diff
@@ -24,14 +24,14 @@
...
-from azurelinuxagent.common.protocol.restapi import Extension, ExtHandler, ExtHandlerVersionUri, VMAgentManifest, \
-    VMAgentManifestUri, ExtensionState, ExtHandlerList, VMAgentManifestList, InVMGoalStateMetaData
+from azurelinuxagent.common.protocol.restapi import Extension, ExtHandler, VMAgentManifest, \
+    ExtensionState, ExtHandlerList, InVMGoalStateMetaData
 from azurelinuxagent.common.utils.textutil import parse_doc, parse_json, findall, find, findtext, getattrib, gettext, format_exception, \
     is_str_none_or_whitespace, is_str_empty
 
 
 class ExtensionsGoalStateFromExtensionsConfig(ExtensionsGoalState):
-    def __init__(self, incarnation, xml_text, wire_client=None):
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py:34]: The default values (None) for wire_client was a left over of a test I was doing during my previous PR.

File: tests/data/hostgaplugin/ext_conf.xml [hunk 2]
```diff
@@ -1,94 +1,119 @@
...
-      <RuntimeSettings seqNo="0">
-        {
-          "runtimeSettings": [
-            {
-              "handlerSettings": {
-                "publicSettings": {"commandToExecute":"echo '09cd27e9-fbd6-48ad-be86-55f3783e0a23'"}
-              }
-            }
-          ]
+<Extensions version="1.0.0.0" goalStateIncarnation="10"><GuestAgentExtension xmlns:i="http://www.w3.org/2001/XMLSchema-instance">
```
Comment (narrieta) [inline tests/data/hostgaplugin/ext_conf.xml:2]: I updated the test data from an actual DCR run to get the latest changes in HostGAPlugin

File: tests/data/hostgaplugin/ext_conf.xml [hunk 3]
```diff
@@ -1,94 +1,119 @@
...
-          "runtimeSettings": [
-            {
-              "handlerSettings": {
-                "publicSettings": {"source":{"script":"echo '737fa9f1-e9bf-4c3e-ab1f-9e03cd0b5b40'"}}
-              }
-            }
-          ]
+    }
+  ]
+}</RuntimeSettings>
```
Comment (narrieta) [inline tests/data/hostgaplugin/ext_conf.xml:62]: I added the extensions with protected settings to this data file and removed the *-protected-settings.* files. This way I do not need to update 2 sets of data files when I want to take an update of the HostGAPlugin schema.

File: azurelinuxagent/common/protocol/extensions_goal_state.py [hunk 4]
```diff
@@ -115,6 +119,7 @@ def compare_attribute(attribute):
             compare_attribute("status_upload_blob_type")
         compare_attribute("required_features")
         compare_attribute("on_hold")
+        compare_attribute("agent_manifests")
```
Comment (larohra) [inline azurelinuxagent/common/protocol/extensions_goal_state.py:122]: Is the comparison of ext-handlers going to be done in a separate PR?

File: azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py [hunk 5]
```diff
@@ -109,8 +139,20 @@ def _parse_simple_attributes(self, vm_settings):
         self._activity_id = self._string_to_id(vm_settings.get("activityId"))
         self._correlation_id = self._string_to_id(vm_settings.get("correlationId"))
         self._created_on_timestamp = self._ticks_to_utc_timestamp(vm_settings.get("extensionsLastModifiedTickCount"))
+        self._source = vm_settings.get("extensionGoalStatesSource")
+        if self._source is None:
+            self._source = "UNKNOWN"
```
Comment (larohra) [inline azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py:144]: Shouldn't this default to Fabric?

File: azurelinuxagent/common/protocol/restapi.py [hunk 6]
```diff
@@ -64,21 +64,22 @@ def __init__(self):
         self.certificates = DataContractList(Cert)
 
 
-# TODO: confirm vmagent manifest schema
-class VMAgentManifestUri(DataContract):
-    def __init__(self, uri=None):
-        self.uri = uri
+class VMAgentManifest(object):
```
Comment (larohra) [inline azurelinuxagent/common/protocol/restapi.py:67]: Thanks for the refactoring!

File: tests/data/hostgaplugin/vm_settings.json [hunk 7]
```diff
@@ -1,21 +1,21 @@
...
-    "correlationId": "1bef4c48-044e-4225-8f42-1d1eac1eb158",
-    "extensionsLastModifiedTickCount": 637693267431616449,
+    "hostGAPluginVersion": "1.0.8.115",
+    "vmSettingsSchemaVersion": "0.0",
+    "activityId": "a33f6f53-43d6-4625-b322-1a39651a00c9",
+    "correlationId": "9a47a2a2-e740-4bfc-b11b-4f2f7cfe7d2e",
+    "extensionsLastModifiedTickCount": 637726657706205217,
     "extensionGoalStatesSource": "Fabric",
-    "StatusUploadBlob": {
+    "statusUploadBlob": {
```
Comment (larohra) [inline tests/data/hostgaplugin/vm_settings.json:8]: I know we handle this on our end, but is the casing not guaranteed from HostGA? I remember Rohit mentioning something about fixing this, is that not in yet?

**Review by larohra — APPROVED** [review]
some minor comments, else LGTM

File: azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py (hunk 5 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py:144]: why? :)

File: azurelinuxagent/common/protocol/extensions_goal_state.py (hunk 4 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state.py:122]: yes, all this refactoring is affecting a lot of code and those changes need to be done gradually

File: tests/data/hostgaplugin/vm_settings.json (hunk 7 above)
Comment (narrieta) [reply tests/data/hostgaplugin/vm_settings.json:8]: we need to keep doing this until all the nodes in the fleet are updated with the newest HostGAPlugin (which will take a loong time, unfortunately)

File: azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py (hunk 5 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py:144]: (we get this from the HostGAPlugin and do not know the source - we can't assume fabric)

Comment (larohra) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_vm_settings.py:144]: haha fair enough :)

File: tests/data/hostgaplugin/vm_settings.json (hunk 7 above)
Comment (larohra) [reply tests/data/hostgaplugin/vm_settings.json:8]: totally understand that, was just curious to know if this was being handled by HostGA or not. Like if its supposed to be handled properly by HostGA and still we're seeing discrepancy we should probably let Rohit know about it too as it might be a regression there

Comment (narrieta) [reply tests/data/hostgaplugin/vm_settings.json:8]: At some point we should be ready to declare "version W.X.Y.Z of the HostGAPlugin is ready and all the fleet is updated"... at that point we'll add a single check for that version and remove all those workarounds/checks (case-insensitive dictionary, check for > 112, comparison of goal states, etc). Discrepancies would pop up then. Right now it is not worth checking the case of properties).

File: azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py [hunk 8]
```diff
@@ -24,14 +24,14 @@
...
 from azurelinuxagent.common.protocol.extensions_goal_state import ExtensionsGoalState
-from azurelinuxagent.common.protocol.restapi import Extension, ExtHandler, ExtHandlerVersionUri, VMAgentManifest, \
-    VMAgentManifestUri, ExtensionState, ExtHandlerList, VMAgentManifestList, InVMGoalStateMetaData
+from azurelinuxagent.common.protocol.restapi import Extension, ExtHandler, VMAgentManifest, \
+    ExtensionState, ExtHandlerList, InVMGoalStateMetaData
 from azurelinuxagent.common.utils.textutil import parse_doc, parse_json, findall, find, findtext, getattrib, gettext, format_exception, \
     is_str_none_or_whitespace, is_str_empty
 
 
 class ExtensionsGoalStateFromExtensionsConfig(ExtensionsGoalState):
```
Comment (nagworld9) [inline azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py:33]: can we test for agent_manifest from extension_config like we have for vm_settings.

Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py:33]: not sure i get your comment @nagworld9 ... the extension config stuff is existing code and has existing tests. is that what you meant?

Comment (nagworld9) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py:33]: Unfortunately, the existing tests not covering agent manifests checks. I don't mind leaving it now but may be when you touch the test file again it's worthwhile to consider testing.
https://github.com/Azure/WALinuxAgent/blob/develop/tests/protocol/test_extensions_goal_state_from_extensions_config.py#L8

**Review by nagworld9 — DISMISSED** [review]
LGTM

File: azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py (hunk 8 above)
Comment (narrieta) [reply azurelinuxagent/common/protocol/extensions_goal_state_from_extensions_config.py:33]: yes, the existing code lacks some tests. retrofitting tests to that code would be costly. do feel free to add tests where you consider it worth the effore

**Review by larohra — APPROVED** [review]

**Review by nagworld9 — APPROVED** [review]
